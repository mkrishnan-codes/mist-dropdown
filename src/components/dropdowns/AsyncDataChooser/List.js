import React, { forwardRef, useState, useCallback, useEffect } from 'react';

const List = forwardRef((props, ref) => {
	const [pos, setPos] = useState(0);
	useEffect(() => {
		if (props.show && pos > 0 && ref.current) {
			ref.current.scrollTo(0, pos)
		}
	}, [props.show]);
	const onScroll = useCallback((e) => {

		setPos(e.currentTarget.scrollTop)
		props.onScroll(e);
	}, []);

	return props.show && (
		<div className={`dropdown-container ${props.showFilter ? `has-filter` : ``}`}>
			{props.showFilter && <div
				className="search-container">
				<input className="search-input" placeholder="Search" value={props.filterValue} onChange={props.onFilterValueChange} />
			</div>

			}
			<div
				className="dropdown"
				style={{ maxHeight: 500 }}
				ref={ref}
				id="async-dropdown-list"
				onScroll={onScroll}>


				<ul style={{ height: props.innerHeight }} >

					{
						props.items.map((item, index) => <li
							style={{ height: props.itemHeight, top: (props.startIndex + index) * props.itemHeight }}
							key={props.keyExtractor(item)}
							datavalue={props.startIndex + index}
							onClick={props.onClick}
						>
							<div className="li-inner">{props.itemRender(item)}</div></li>)
					}

				</ul>
			</div>
		</div>
	);
})

export default List;
