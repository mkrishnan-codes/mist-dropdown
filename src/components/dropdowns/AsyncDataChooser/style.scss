.ac-dropdown {
    margin: 0;
    position: relative;
    cursor: pointer;
    display: flex;
    flex-direction: row;
    border: 1px solid #d2d2d7;
    padding: 10px;
    background-color: #ffffff;

    .dropdown-container {
        position: absolute;
        margin-top: 1px;
        top: 100%;
        width: 100%;
        left: 0px;

        box-shadow: 0 0 6px 0 rgba(0, 0, 0, 0.1);

        border: 1px solid #d2d2d7;
        border-radius: 4px;
        background-color: #ffffff;

        height: 500px;

        &.has-filter {
            height: 550px;

            .dropdown {
                top: 50px;
            }
        }

        .search-container {
            display: flex;
            justify-content: center;
            align-items: center;
            height: 50px;
            padding: 10px;

            .search-input {
                width: 100%;
                border: 1px solid #d2d2d7;
                padding: 5px;

                &:focus {
                    outline: none;
                }
            }
        }
    }

    .test-height {
        background: red;
        margin: 0px;
        padding: 0px;
        list-style: none;
        position: absolute;
        width: 100%;
        visibility: hidden;
    }

    .arrow {
        // margin-left: 5px;
        border: solid #333;
        border-width: 0 2px 2px 0;
        display: inline-block;
        padding: 2px;

        &.down {
            transform: rotate(-135deg);
            -webkit-transform: rotate(-135deg);
        }

        &.up {
            transform: rotate(45deg);
            -webkit-transform: rotate(45deg);
        }
    }

    .dropdown-button {
        display: flex;
        flex-direction: row;
        justify-content: space-between;
        align-items: center;
        width: 100%;

        .truncated {
            flex: 1;

            white-space: nowrap;
            overflow: hidden;
            text-overflow: ellipsis;
        }

        &:focus {
            outline: none;
        }
    }

    .dropdown {
        position: absolute;
        left: 0;
        top: 0;

        padding: 0;
        margin: 2px 0 0 0;
        overflow-y: scroll;
        width: 100%;



        ul {
            margin: 0px;
            padding: 0px;

            li {
                list-style: none;
                position: absolute;
                width: 100%;
                left: 0px;

                .li-inner {
                    padding: 5px;
                }
            }


        }
    }
}

.loader-container {
    position: absolute;
    top: 100%;
    width: 100%;
    height: 100px;
    left: 0;
    display: flex;
    margin-top: 1px;
    justify-content: center;
    align-items: center;
    box-shadow: 0 0 6px 0 rgba(0, 0, 0, 0.1);

    border: 1px solid #d2d2d7;
    border-radius: 4px;
    background-color: #ffffff;
}

