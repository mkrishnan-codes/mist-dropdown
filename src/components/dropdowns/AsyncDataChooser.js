import React, { useState, useEffect } from 'react';
import './style.scss';

const getData = async () => {
	const airports = await import('./configs/airports.js');
	return airports['default'];
};
const AsyncDataChooser = () => {
	const [show, setShow] = useState(false);
	const [data, setdata] = useState([]);
	const [scrollTop, setScrollTop] = useState(0);
	const numItems = data.length;
	const itemHeight = 40;
	const windowHeight = 500;
	useEffect(() => {
		const callApi = async () => {
			const dt = await getData();
			setdata(dt);
		};
		callApi();
	}, []);
	const innerHeight = numItems * itemHeight;
	const startIndex = Math.floor(scrollTop / itemHeight);
	const endIndex = Math.min(
		numItems - 1, 
		Math.floor((scrollTop + windowHeight) / itemHeight)
	);
	const items = [];
	const setItem = (it, _ind) => {
		it._ind = _ind;
		return it;
	}
	for (let i = startIndex; i <= endIndex; i++) {
		items.push(
			setItem(data[i], i)
		);
	}
	const onScroll = e => setScrollTop(e.currentTarget.scrollTop);
	return (
		<div className="ac-dropdown">
			<span onClick={() => setShow(!show)} className="selected"><span>Dropdown</span>
				<i className={`arrow ${show ? `down` : 'up'}`}></i>
			</span>
			{show && <div className="dropdown" style={{  maxHeight: 500 }} onScroll={onScroll}>
				<ul style={{ height: innerHeight }} >
					{
						items.map((item) => <li style={{ height: itemHeight, position: 'absolute', top: item._ind * itemHeight, width:'100%' }} key={item.code}>{item.code}</li>)
					}
				</ul>
			</div>}
		</div>
	);
}

export default AsyncDataChooser;
