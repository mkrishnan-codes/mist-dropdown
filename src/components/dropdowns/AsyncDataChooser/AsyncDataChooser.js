import React, { useState, useEffect, useLayoutEffect, useRef, useCallback } from 'react';
import PropTypes from 'prop-types';
import './style.scss';
import List from './List';
// const dataReducer = (state, action) => {
// 	switch (action.type) {
// 		case 'ADD':
// 			{
// 				return action.payload
// 			}
// 		default: return state;
// 	}
// }

const AsyncDataChooser = (props) => {
	let testDivRef = useRef(null);
	let containerRef = useRef(null);
	let ulContainerRef = useRef(null);
	const [show, setShow] = useState(false);
	const [hieghtCalculated, setHieghtCalculated] = useState(false);
	const [data, setData] = useState([]);
	const [itemHeight, setItemHeight] = useState(100);
	const [scrollTop, setScrollTop] = useState(0);
	const numItems = data.length;
	const windowHeight = 500;
	useEffect(() => {
		const callApi = async () => {
			const dt = await props.getAsyncData();
			setData(dt);
		};
		callApi();
	}, []);


	const innerHeight = numItems * itemHeight;
	const startIndex = Math.floor(scrollTop / itemHeight);
	const endIndex = Math.min(
		numItems - 1,
		Math.floor((scrollTop + windowHeight) / itemHeight)
	);
	const items = [];
	for (let i = startIndex; i <= endIndex; i++) {
		items.push(data[i]);
	}
	const renderTest = items.length > 0;
	useLayoutEffect(() => {
		// const rect = testDivRef.current && testDivRef.current.getBoundingClientRect();
		// console.log(rect, 'RECT');
		if (!hieghtCalculated && testDivRef.current) {
			const height = testDivRef.current.clientHeight
			setItemHeight(height)
			setHieghtCalculated(true)
		}
	}, [renderTest])
	const handleClickOutside = (event) => {
		if (containerRef.current && !containerRef.current.contains(event.target)) {
			setShow(false)
		}
	}
	useEffect(() => {
		document.addEventListener("mousedown", handleClickOutside);
		return () => {
			document.removeEventListener("mousedown", handleClickOutside);
		};
	}, []);
	const onScroll = e => setScrollTop(e.currentTarget.scrollTop);
	const onClick = useCallback((e) => {
		setShow(false);
		if (props.onSelect) {
			props.onSelect(data[e.currentTarget.getAttribute('datavalue')])
		}
	})

	return (
		<div
			style={props.style}
			ref={containerRef}
			className={`ac-dropdown ${props.className ? props.className : ``}`} >
			<div
				role="button"
				aria-label={props.label}
				tabIndex={0}
				onClick={() => setShow(!show)}
				className="dropdown-button">
				<div className="truncated">{props.label}</div>
				<div className={`arrow ${show ? `down` : 'up'}`}></div>
			</div>
			<List show={show}
				items={items}
				innerHeight={innerHeight}
				startIndex={startIndex}
				onScroll={onScroll}
				onClick={onClick}
				ref={ulContainerRef}
				startIndex={startIndex}
				keyExtractor={props.keyExtractor}
				itemRender={props.itemRender}
				itemHeight={itemHeight}
			/>
			{renderTest && <li ref={testDivRef} className="test-height" key="hidden">{props.itemRender(items[0])}</li>}
		</div>
	);
}
AsyncDataChooser.defaultProps = {
	label: 'Choose'
}
AsyncDataChooser.propTypes = {
	getAsyncData: PropTypes.func.isRequired,
	itemRender: PropTypes.func.isRequired,
	buttonRender: PropTypes.func,
	keyExtractor: PropTypes.func.isRequired,
	label: PropTypes.string,
	style: PropTypes.object,
	onChange: PropTypes.func,
	className: PropTypes.string,
	keepScrollPosition: PropTypes.bool,
}
export default AsyncDataChooser;
