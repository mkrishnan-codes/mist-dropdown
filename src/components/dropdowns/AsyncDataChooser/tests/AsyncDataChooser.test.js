import React from 'react';
import AsyncDataChooser from '../AsyncDataChooser';
import renderer from 'react-test-renderer';
import { shallow, configure } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
configure({ adapter: new Adapter() });
const mockData = async () => {
	return ['One', 'Two', 'Three'];
};
const mockFourData = async () => {
	return ['One', 'Two', 'Three', 'four'];
};
test('Snapshot test', () => {
	const component = renderer.create(
		<AsyncDataChooser
			className="airport-chooser"
			getAsyncData={mockData}
			keyExtractor={item => item.code}
			itemRender={item => item}
			label={'Choose airport'}
			onSelect={item => item}
		/>,
	);
	let tree = component.toJSON();
	expect(tree).toMatchSnapshot();
});

test('Render Button with default label', () => {
	const dataChoser = shallow(<AsyncDataChooser
		className="airport-chooser"
		getAsyncData={mockData}
		keyExtractor={item => item.code}
		itemRender={item => item}
		onSelect={item => item}
	/>);

	expect(dataChoser.text()).toEqual('Choose');
});
test('Render Button with specified label', () => {
	const dataChoser = shallow(<AsyncDataChooser
		className="airport-chooser"
		getAsyncData={mockData}
		keyExtractor={item => item.code}
		itemRender={item => item}
		label={'Select item'}
		onSelect={item => item}
	/>);

	expect(dataChoser.text()).toEqual('Select item');
});
describe('Test list show and hide', () => {
	it('Check list is not loaded first', () => {
		const wrapper = shallow((<AsyncDataChooser
			className="airport-chooser"
			getAsyncData={mockData}
			keyExtractor={item => item.code}
			itemRender={item => item}
			label={'Select item'}
			onSelect={item => item}
		/>));
		expect(wrapper.containsMatchingElement(<ul />)).toEqual(false);
	});
	it('Check list is present after click', () => {
		const wrapper = shallow((<AsyncDataChooser
			className="airport-chooser"
			getAsyncData={mockData}
			keyExtractor={item => item.code}
			itemRender={item => item}
			label={'Select item'}
			onSelect={item => item}
		/>));

		wrapper.find('.dropdown-button').simulate('click');
		setTimeout(() => {
			expect(wrapper.containsMatchingElement(<ul />)).toEqual(true);
		}, 100);

	});
	it('Check list item length with 4 items', () => {
		const wrapper = shallow((<AsyncDataChooser
			className="airport-chooser"
			getAsyncData={mockFourData}
			keyExtractor={item => item.code}
			itemRender={item => item}
			label={'Select item'}
			onSelect={item => item}
		/>));

		wrapper.find('.dropdown-button').simulate('click');
		setTimeout(() => {
			expect(wrapper.find('ul').children().length).toEqual(4);
		}, 100);

	});
	it('Check list item length with 4 items', () => {
		const wrapper = shallow((<AsyncDataChooser
			className="airport-chooser"
			getAsyncData={mockData.push}
			keyExtractor={item => item.code}
			itemRender={item => item}
			label={'Select item'}
			onSelect={item => item}
		/>));

		wrapper.find('.dropdown-button').simulate('click');
		setTimeout(() => {
			expect(wrapper.find('ul').children().length).toEqual(3);
		}, 100);

	});

});