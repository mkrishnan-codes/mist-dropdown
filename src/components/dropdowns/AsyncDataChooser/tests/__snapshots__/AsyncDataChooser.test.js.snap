// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Snapshot test 1`] = `
<div
  className="ac-dropdown airport-chooser"
>
  <div
    aria-label="Choose airport"
    className="dropdown-button"
    onClick={[Function]}
    role="button"
    tabIndex={0}
  >
    <div
      className="truncated"
    >
      Choose airport
    </div>
    <div
      className="arrow up"
    />
  </div>
</div>
`;
